ESP32 board and nRF52 radio compatibility

mysensors has got ESP32 support by tekka007, I was assuming, that also Nodemanager will be ESP32 supported, but it's maintainted seperately.
Since I'm using mysensors with ESP32 barebone boards (for the gateway and also for the sensor nodes), it would be nice to use Nodemanager with ESP32 otherwise I can't use it and would do my own implementation based on the mysensors core.
Since ESP8266 seem to be already supported, I assume, that the ESP32 support should be doable.
I have started to do it by myself, but it would be more efficient, if somebody who knows the mysensors core and nodemanager much better will do it.

@user2684 I took the Development branch (is that what you want me to test)?
So I took the Development branch and coiped it to the libraries and since my last working ESP32 Nodemanger does no longer work, I took your exaples/template.ino and just seleted a few settings like the RFM95 and the battery, analogread and the signal features (uncommended the lines)

However I get a lot of compile errors. I also took the latest MySensors Development libraries afterwards because I though my existing ones (2.3 release) does no longer match. But the same result.
I think I had such a list of errors already but I can't remember what I did. I thought is again the #undef CHIP_AVR I had to add, but it did not solve it. I guess there preprosessore still think I have an AVR chip and not an ESP32.
So if you're saying you have compiled it with an chosen ESP32 board (not having one connected), what board have you chosen? I have tested at least 4 different ESP32 boards, although I have a bare ESP-WROOM-32, also called ESP32S mounted on a Mysensors board from ranseyer). It has already compiled and did run with my adapted code long time ago. 

Compile errors:
In file included from D:\mcdev\arduino\portable\sketchbook\libraries\NodeManager/MySensors_NodeManager.h:82:0,
                 from D:\mcdev\arduino\portable\sketchbook\_ESP32\ESP32_NodeManager_Template\ESP32_NodeManager_Template.ino:178:
D:\mcdev\arduino\portable\sketchbook\libraries\NodeManager/nodemanager/Node.cpp: In member function 'void NodeManager::reboot()':
D:\mcdev\arduino\portable\sketchbook\libraries\NodeManager/nodemanager/Node.cpp:356:4: error: 'WDTCSR' was not declared in this scope
    WDTCSR |= (1<<WDCE) | (1<<WDE);
    ^
D:\mcdev\arduino\portable\sketchbook\libraries\NodeManager/nodemanager/Node.cpp:356:18: error: 'WDCE' was not declared in this scope
    WDTCSR |= (1<<WDCE) | (1<<WDE);
                  ^
D:\mcdev\arduino\portable\sketchbook\libraries\NodeManager/nodemanager/Node.cpp:356:30: error: 'WDE' was not declared in this scope
    WDTCSR |= (1<<WDCE) | (1<<WDE);
                              ^
D:\mcdev\arduino\portable\sketchbook\libraries\NodeManager/nodemanager/Node.cpp: In member function 'void NodeManager::setADCOff()':
D:\mcdev\arduino\portable\sketchbook\libraries\NodeManager/nodemanager/Node.cpp:449:3: error: 'ADCSRA' was not declared in this scope
   ADCSRA = ADCSRA & B01111111;
   ^
D:\mcdev\arduino\portable\sketchbook\libraries\NodeManager/nodemanager/Node.cpp:451:3: error: 'ACSR' was not declared in this scope
   ACSR = B10000000;
   ^
D:\mcdev\arduino\portable\sketchbook\libraries\NodeManager/nodemanager/Node.cpp: In member function 'void NodeManager::sleepOrWait(long unsigned int)':
D:\mcdev\arduino\portable\sketchbook\libraries\NodeManager/nodemanager/Node.cpp:458:69: error: call of overloaded 'sleep(long unsigned int&)' is ambiguous
   if (isSleepingNode() && _sleep_or_wait && value > 200) sleep(value);
                                                                    ^
In file included from D:\mcdev\arduino\portable\sketchbook\libraries\MySensors/MySensors.h:433:0,
                 from D:\mcdev\arduino\portable\sketchbook\libraries\NodeManager/MySensors_NodeManager.h:43,
                 from D:\mcdev\arduino\portable\sketchbook\_ESP32\ESP32_NodeManager_Template\ESP32_NodeManager_Template.ino:178:
D:\mcdev\arduino\portable\sketchbook\libraries\MySensors/core/MySensorsCore.cpp:666:8: note: candidate: int8_t sleep(uint32_t, bool)
 int8_t sleep(const uint32_t sleepingMS, const bool smartSleep)
        ^
In file included from D:\mcdev\arduino\portable\sketchbook\hardware\espressif\esp32/tools/sdk/include/newlib/stdio.h:29:0,
                 from D:\mcdev\arduino\portable\sketchbook\hardware\espressif\esp32\cores\esp32/Arduino.h:27,
                 from sketch\ESP32_NodeManager_Template.ino.cpp:1:
D:\mcdev\arduino\portable\sketchbook\hardware\espressif\esp32/tools/sdk/include/newlib/sys/unistd.h:185:17: note: candidate: unsigned int sleep(unsigned int)
 unsigned _EXFUN(sleep, (unsigned int __seconds ));
                 ^
D:\mcdev\arduino\portable\sketchbook\hardware\espressif\esp32/tools/sdk/include/newlib/_ansi.h:65:30: note: in definition of macro '_EXFUN'
 #define _EXFUN(name, proto)  name proto
                              ^
In file included from D:\mcdev\arduino\portable\sketchbook\libraries\NodeManager/MySensors_NodeManager.h:82:0,
                 from D:\mcdev\arduino\portable\sketchbook\_ESP32\ESP32_NodeManager_Template\ESP32_NodeManager_Template.ino:178:
D:\mcdev\arduino\portable\sketchbook\libraries\NodeManager/nodemanager/Node.cpp: In member function 'void NodeManager::setAnalogReference(uint8_t, uint8_t)':
D:\mcdev\arduino\portable\sketchbook\libraries\NodeManager/nodemanager/Node.cpp:731:23: error: 'analogReference' was not declared in this scope
  analogReference(value);
                       ^
In file included from D:\mcdev\arduino\portable\sketchbook\_ESP32\ESP32_NodeManager_Template\ESP32_NodeManager_Template.ino:186:0:
D:\mcdev\arduino\portable\sketchbook\libraries\NodeManager/sensors/SensorBattery.h: In member function 'virtual void SensorBattery::onLoop(Child*)':
D:\mcdev\arduino\portable\sketchbook\libraries\NodeManager/sensors/SensorBattery.h:85:35: error: 'INTERNAL' was not declared in this scope
    nodeManager.setAnalogReference(INTERNAL);
                                   ^

@user2684 ok, I did not know how to get the branch for PR #414 (still don't know how to link the pull request # to the branch) but with your link to the branch I did the dowaload, installed the lib and took the template example from this branch and still get compile errors (but less). Obviously something with the sleep() does not match (with the mysensors library)?

In file included from D:\mcdev\arduino\portable\sketchbook\libraries\NodeManager/MySensors_NodeManager.h:82:0,
                 from D:\mcdev\arduino\portable\sketchbook\_ESP32\ESP32_NodeManager_Template\ESP32_NodeManager_Template.ino:176:
D:\mcdev\arduino\portable\sketchbook\libraries\NodeManager/nodemanager/Node.cpp: In member function 'void NodeManager::sleepOrWait(long unsigned int)':
D:\mcdev\arduino\portable\sketchbook\libraries\NodeManager/nodemanager/Node.cpp:458:69: error: call of overloaded 'sleep(long unsigned int&)' is ambiguous
   if (isSleepingNode() && _sleep_or_wait && value > 200) sleep(value);
                                                                     ^
In file included from D:\mcdev\arduino\portable\sketchbook\libraries\MySensors/MySensors.h:433:0,
                 from D:\mcdev\arduino\portable\sketchbook\libraries\NodeManager/MySensors_NodeManager.h:43,
                 from D:\mcdev\arduino\portable\sketchbook\_ESP32\ESP32_NodeManager_Template\ESP32_NodeManager_Template.ino:176:
D:\mcdev\arduino\portable\sketchbook\libraries\MySensors/core/MySensorsCore.cpp:666:8: note: candidate: int8_t sleep(uint32_t, bool)
 int8_t sleep(const uint32_t sleepingMS, const bool smartSleep)
        ^
In file included from D:\mcdev\arduino\portable\sketchbook\hardware\espressif\esp32/tools/sdk/include/newlib/stdio.h:29:0,
                 from D:\mcdev\arduino\portable\sketchbook\hardware\espressif\esp32\cores\esp32/Arduino.h:27,
                 from sketch\ESP32_NodeManager_Template.ino.cpp:1:
D:\mcdev\arduino\portable\sketchbook\hardware\espressif\esp32/tools/sdk/include/newlib/sys/unistd.h:185:17: note: candidate: unsigned int sleep(unsigned int)
 unsigned _EXFUN(sleep, (unsigned int __seconds ));
                 ^
D:\mcdev\arduino\portable\sketchbook\hardware\espressif\esp32/tools/sdk/include/newlib/_ansi.h:65:30: note: in definition of macro '_EXFUN'
 #define _EXFUN(name, proto)  name proto

BTW: I knwo from MySensors, that the ESP32 ist supported and does not show compile errors, but the sleep does not stop/sleep but just continues an returns an error code (not supported) which can be ignored.
Fot MySensors Gateways, the ESP32 works perfect, even ESP32 with RFM95 on-board. I use it with that configuration.
But it would be nice, if Nodemanager would compile and work without sleep for all other use cases.

What I'm now trying is to get Mysensors/Node Manager running on my nRF52 boards which are really fit 100% to battery powered projects. Lets see what errors I get then (nRF52 should be supportd in principle) and have the expected sleep/wakeup behavior (keeps its state, does not a restart etc.)

@user2684 Now to the secon d interesting topic, the support for the nRF52832:
I have trried a few things like the board manager extension from sandeep mistry
This is what you have to include in the Arduino Preferneces - new board manager URL's:
https://sandeepmistry.github.io/arduino-nRF5/package_nRF5_boards_index.json
And then you have to restart Arduino and go to Tools - Board Manager and it loads the new boards wheer you can pick one e.g. Generic nRF5, but non of them help really (you get a lot of compile errors in MySensors/NodeManger).
Then I got the hint to use this board extension made for MySensors:
https://raw.githubusercontent.com/mysensors/ArduinoBoards/master/package_mysensors.org_index.json
Some procedure as above but you get additional two boards (MyNRF5 nRF52832 ...)
Choose the nRF52832 board.
https://raw.githubusercontent.com/mysensors/ArduinoBoards/master/package_mysensors.org_index.json 

When I then try to compile the Nodemanager I get the following errors:

in file included from D:\mcdev\arduino\portable\sketchbook\libraries\NodeManager/MySensors_NodeManager.h:82:0,
                 from D:\mcdev\arduino\portable\sketchbook\_nRF52\nRF52_NodeManager_Template\nRF52_NodeManager_Template.ino:176:

D:\mcdev\arduino\portable\sketchbook\libraries\NodeManager/nodemanager/Node.cpp: In member function 'void NodeManager::sendMessage(uint8_t, uint8_t, int)':

D:\mcdev\arduino\portable\sketchbook\libraries\NodeManager/nodemanager/Node.cpp:529:21: error: call of overloaded 'set(int&)' is ambiguous
   _message.set(value);
                     ^

In file included from D:\mcdev\arduino\portable\sketchbook\libraries\MySensors/MySensors.h:431:0,
                 from D:\mcdev\arduino\portable\sketchbook\libraries\NodeManager/MySensors_NodeManager.h:43,

                 from D:\mcdev\arduino\portable\sketchbook\_nRF52\nRF52_NodeManager_Template\nRF52_NodeManager_Template.ino:176:

D:\mcdev\arduino\portable\sketchbook\libraries\MySensors/core/MyMessage.cpp:268:12: note: candidate: MyMessage& MyMessage::set(bool)

 MyMessage& MyMessage::set(const bool value)
            ^

D:\mcdev\arduino\portable\sketchbook\libraries\MySensors/core/MyMessage.cpp:276:12: note: candidate: MyMessage& MyMessage::set(uint8_t)

 MyMessage& MyMessage::set(const uint8_t value)

            ^

D:\mcdev\arduino\portable\sketchbook\libraries\MySensors/core/MyMessage.cpp:293:12: note: candidate: MyMessage& MyMessage::set(uint32_t)

 MyMessage& MyMessage::set(const uint32_t value)

            ^

D:\mcdev\arduino\portable\sketchbook\libraries\MySensors/core/MyMessage.cpp:301:12: note: candidate: MyMessage& MyMessage::set(int32_t)

 MyMessage& MyMessage::set(const int32_t value)

            ^

D:\mcdev\arduino\portable\sketchbook\libraries\MySensors/core/MyMessage.cpp:309:12: note: candidate: MyMessage& MyMessage::set(uint16_t)

 MyMessage& MyMessage::set(const uint16_t value)

            ^

D:\mcdev\arduino\portable\sketchbook\libraries\MySensors/core/MyMessage.cpp:317:12: note: candidate: MyMessage& MyMessage::set(int16_t)

 MyMessage& MyMessage::set(const int16_t value)

            ^

--- a/nodemanager/Constants.h
+++ b/nodemanager/Constants.h
@@ -86,10 +86,13 @@ Chip type
 #if defined(ESP8266) || defined(MY_GATEWAY_ESP8266)
 #define CHIP_ESP8266
 #endif
+#if defined(ESP32) || defined(MY_GATEWAY_ESP32)
+#define CHIP_ESP32
+#endif
 #if defined (MYBOARDNRF5)
 #define CHIP_NRF5
 #endif
-#if !defined(CHIP_ESP8266) && !defined(CHIP_STM32) && !defined(CHIP_NRF5)
+#if !defined(CHIP_ESP8266) && !defined(CHIP_ESP32) && !defined(CHIP_STM32) && !defined(CHIP_NRF5)
 #define CHIP_AVR
 #endif
 

--- a/nodemanager/Node.cpp
+++ b/nodemanager/Node.cpp
@@ -526,7 +526,7 @@ void NodeManager::loop() {
 	// send a message by providing the source child, type of the message and value
 	void NodeManager::sendMessage(uint8_t child_id, uint8_t type, int value) {
 		_message.clear();
-		_message.set(value);
+		_message.set((int32_t)value);
 		_sendMessage(child_id,type);
 	}
 	void NodeManager::sendMessage(uint8_t child_id, uint8_t type, float value, uint8_t precision) {

--- a/nodemanager/Node.cpp
+++ b/nodemanager/Node.cpp
@@ -455,7 +455,7 @@ void NodeManager::loop() {
 	// sleep if the node is a battery powered or wait if it is not for the given number of milliseconds 
 	void NodeManager::sleepOrWait(unsigned long value) {
 		// if the node is sleeping, sleep-or-wait is enabled and we need to sleep for a decent amount of time, call sleep() otherwise wait()
-		if (isSleepingNode() && _sleep_or_wait && value > 200) sleep(value);
+		if (isSleepingNode() && _sleep_or_wait && value > 200) sleep(value,false);
 		else wait(value);
 	}
 

